pkg/commands/helpers/kubectl_auth.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// InitDefaults populates the default values of options
+func (o *HelperKubectlAuthOptions) InitDefaults() {
+	o.Duration = 1 * time.Hour
+}
+
+// NewCmdHelperKubectlAuth builds a cobra command for the kubectl-auth command
+func NewCmdHelperKubectlAuth(f *util.Factory, out io.Writer) *cobra.Command {
+	options := &HelperKubectlAuthOptions{}
+	options.InitDefaults()
+
+	cmd := &cobra.Command{
+		Use:   "kubectl-auth",
+		Short: kubectlAuthShort,
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx := context.TODO()
+
+			err := RunKubectlAuthHelper(ctx, f, out, options)
+			if err != nil {
+				commandutils.ExitWithError(err)
+			}
+		},
+	}
+
+	cmd.Flags().StringVar(&options.APIVersion, "api-version", options.APIVersion, "version of client.authentication.k8s.io schema in use")
+	cmd.Flags().StringVar(&options.ClusterName, "cluster", options.ClusterName, "cluster to target")
+	cmd.Flags().DurationVar(&options.Duration, "validity", options.Duration, "attempt to issue a credential with this validity")
+
+	return cmd
+}
+
+// RunKubectlAuthHelper implements the kubectl auth helper, which creates an authentication token
+func RunKubectlAuthHelper(ctx context.Context, f *util.Factory, out io.Writer, options *HelperKubectlAuthOptions) error {
+	if options.ClusterName == "" {
+		return fmt.Errorf("ClusterName is required")
+	}
+
+	execCredential := &ExecCredential{
+		Kind: "ExecCredential",
+	}
+
+	switch options.APIVersion {
+	case "":
+		return fmt.Errorf("api-version must be specified")
+	case "v1alpha1":
+		execCredential.APIVersion = "client.authentication.k8s.io/v1alpha1"
+	case "v1beta1":
+		execCredential.APIVersion = "client.authentication.k8s.io/v1beta1"
+
+	default:
+		return fmt.Errorf("api-version %q is not supported", options.APIVersion)
+	}
+
+	cacheFilePath := cacheFilePath(f.KopsStateStore(), options.ClusterName)
+	cached, err := loadCachedExecCredential(cacheFilePath)
+	if err != nil {
+		klog.Warningf("cached credential %q was not valid: %v", cacheFilePath, err)
+		cached = nil
+	}
+
+	if cached != nil && cached.APIVersion != execCredential.APIVersion {
+		klog.Warningf("cached credential had wrong api version")
+		cached = nil
+	}
+
+	isCached := false
+	if cached != nil {
+		execCredential = cached
+		isCached = true
+	} else {
+		status, err := buildCredentials(ctx, f, options)
+		if err != nil {
+			return err
+		}
+		execCredential.Status = *status
+	}
+
+	b, err := json.MarshalIndent(execCredential, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error marshaling json: %v", err)
+	}
+	_, err = out.Write(b)
+	if err != nil {
+		return fmt.Errorf("error writing to stdout: %v", err)
+	}
+
+	if !isCached {
+		if err := os.MkdirAll(filepath.Dir(cacheFilePath), 0755); err != nil {
+			klog.Warningf("failed to make cache directory for %q: %v", cacheFilePath, err)
+		}
+		if err := ioutil.WriteFile(cacheFilePath, b, 0600); err != nil {
+			klog.Warningf("failed to write cache file %q: %v", cacheFilePath, err)
+		}
```

Comment:
Ah, so you have to cache these in a separate file. Our authentication plugin tries to cache them in the kubecfg, which seems to result in failure whenever it has to update the token.

Reply:
Yes.  I thought it was safer to put them in another file - one of the things I don't like about the GCE plugin is that your token ends up in your kubeconfig and is easily accidentally shared.  Putting them in another file also means that (in theory) we could also store them in a more secure location (e.g. https://github.com/zalando/go-keyring).  But it is another file which means a bigger surface we now have to think about.

How does your authentication plugin write back to the kubeconfig?  Does it use the exec plugin mechanism, and is there a trick I'm missing?

Reply:
It uses `k8s.io/client-go/tools/clientcmd/api/ModifyConfig()` to write the new token and expiration to the kubeconfig. That unfortunately doesn't affect the current kubectl process, so the first kubectl command usually fails after updating the kubecfg and the next one works.